### PR TITLE
Add Sql Collector (Part 2)

### PR DESCRIPTION
**Pull request by CodeBlanch**
Picking up where @alexvaluyskiy left off on #355 adding a SqlCollector to OpenTelemetry.Collector.Dependencies.

### CODE REVIEW

**Comment by CodeBlanch** [comment]
I tried to pick up all the remaining feedback on #355. The only thing I didn't do was rename SqlCollector->SqlAdapter. It just seemed weird having all the other stuff being mismatched. I figured it would be easier to do everything in one shot while working on #335, whenever that happens.

File: src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs [hunk 1]
```diff
@@ -30,12 +30,12 @@ internal static class ZipkinConversionExtensions
 
         private static readonly Dictionary<string, int> RemoteEndpointServiceNameKeyResolutionDictionary = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
         {
-            ["net.peer.name"] = 0, // RemoteEndpoint.ServiceName primary.
-            ["peer.service"] = 0, // RemoteEndpoint.ServiceName primary.
-            ["peer.hostname"] = 1, // RemoteEndpoint.ServiceName alternative.
-            ["peer.address"] = 1, // RemoteEndpoint.ServiceName alternative.
-            ["http.host"] = 2, // RemoteEndpoint.ServiceName for Http.
-            ["db.instance"] = 2, // RemoteEndpoint.ServiceName for Redis.
+            [SpanAttributeConstants.PeerServiceKey] = 0, // RemoteEndpoint.ServiceName primary.
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs]: I made Zipkin & Jaeger consistent, but I wasn't sure what should have top priority: "net.peer.name" or "peer.service." I went with "peer.service" because that's what it is called in the Jaeger protocol, but OT spec calls out "net.peer.name" specifically. I have a feeling it will change a bunch.

File: src/OpenTelemetry.Api/Trace/SpanExtensions.cs [hunk 2]
```diff
@@ -208,11 +221,50 @@ public static TelemetrySpan PutHttpStatusCode(this TelemetrySpan span, int statu
...
+        /// <param name="type">Database type.</param>
+        /// <returns>Span with populated properties.</returns>
+        public static TelemetrySpan PutDatabaseTypeAttribute(this TelemetrySpan span, string type)
+        {
+            span.SetAttribute(SpanAttributeConstants.DatabaseTypeKey, type);
+            return span;
+        }
+
+        /// <summary>
+        /// Helper method that populates database type
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Trace/SpanExtensions.cs]:
```suggestion
        /// Helper method that populates database instance
```

File: src/OpenTelemetry.Api/Trace/SpanExtensions.cs [hunk 3]
```diff
@@ -208,11 +221,50 @@ public static TelemetrySpan PutHttpStatusCode(this TelemetrySpan span, int statu
...
+            span.SetAttribute(SpanAttributeConstants.DatabaseInstanceKey, instance);
+            return span;
+        }
+
+        /// <summary>
+        /// Helper method that populates database statement
+        /// to https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-database.md.
+        /// </summary>
+        /// <param name="span">Span to fill out.</param>
+        /// <param name="statement">Database type.</param>
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Trace/SpanExtensions.cs]:
```suggestion
        /// <param name="statement">Database statement.</param>
```

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            switch (name)
+            {
+                case SqlDataBeforeExecuteCommand:
+                case SqlMicrosoftBeforeExecuteCommand:
+                    {
+                        var command = this.commandFetcher.Fetch(payload);
+
+                        if (command == null)
+                        {
+                            CollectorEventSource.Log.NullPayload(nameof(SqlClientDiagnosticListener) + name);
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: Need separator between `SqlClientDiagnosticListener` and `name`. Current structure would look like `SqlClientDiagnosticListenerSystem.Data.SqlClient.WriteCommandBefore`, the suggestion output would be `SqlClientDiagnosticListenerSystem-System.Data.SqlClient.WriteCommandBefore`
```suggestion
                            CollectorEventSource.Log.NullPayload($"{nameof(SqlClientDiagnosticListener)}-{name}");
```

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+                    break;
+                case SqlDataAfterExecuteCommand:
+                case SqlMicrosoftAfterExecuteCommand:
+                    {
+                        var span = this.Tracer.CurrentSpan;
+
+                        if (span == null || !span.Context.IsValid)
+                        {
+                            CollectorEventSource.Log.NullOrBlankSpan(nameof(SqlClientDiagnosticListener) + name);
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: Same as above
```suggestion
                            CollectorEventSource.Log.NullOrBlankSpan($"{nameof(SqlClientDiagnosticListener)}-{name}");
```

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+                    break;
+                case SqlDataWriteCommandError:
+                case SqlMicrosoftWriteCommandError:
+                    {
+                        var span = this.Tracer.CurrentSpan;
+
+                        if (span == null || !span.Context.IsValid)
+                        {
+                            CollectorEventSource.Log.NullOrBlankSpan(nameof(SqlClientDiagnosticListener) + name);
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]:
```suggestion
                            CollectorEventSource.Log.NullOrBlankSpan($"{nameof(SqlClientDiagnosticListener)}-{name}");
```

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                            return;
+                        }
+
+                        if (span.IsRecording)
+                        {
+                            if (!(this.exceptionFetcher.Fetch(payload) is Exception exception))
+                            {
+                                CollectorEventSource.Log.NullPayload(nameof(SqlClientDiagnosticListener) + name);
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]:
```suggestion
                                CollectorEventSource.Log.NullPayload($"{nameof(SqlClientDiagnosticListener)}-{name}");
```

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+                        if (span == null || !span.Context.IsValid)
+                        {
+                            CollectorEventSource.Log.NullOrBlankSpan(nameof(SqlClientDiagnosticListener) + name);
+                            return;
+                        }
+
+                        if (span.IsRecording)
+                        {
+                            if (!(this.exceptionFetcher.Fetch(payload) is Exception exception))
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: I think this would be easier to understand if an if/else was used instead of inverting the `is` check.

```csharp
if (this.exceptionFetch.Fetch(payload) is Exception exception)
{
    span.Status = Status.Unknown.WithDescription(exception.Message);
}
else
{
	CollectorEventSource.Log.NullPayload(nameof(SqlClientDiagnosticListener) + name);
}

File: src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs (hunk 1 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs]: This isn't related to the SqlCollector - please can you move these changes into a separate PR where we can discuss what the priority should be too.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs [hunk 9]
```diff
@@ -47,7 +47,7 @@ internal static class JaegerConversionExtensions
 
         private static readonly Dictionary<string, int> PeerServiceKeyResolutionDictionary = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
         {
-            ["peer.service"] = 0, // peer.service primary.
+            [SpanAttributeConstants.PeerServiceKey] = 0, // peer.service primary.
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs]: Please move to separate PR.

File: src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs]: @MikeGoldsmith Would you be open to leaving on here and then I'll open an issue ticket for further discussion? Just trying to save a little time.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs (hunk 9 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs]: All I did here was move "peer.service" to a constant so I could check it in the unit test(s).

File: test/OpenTelemetry.Collector.Dependencies.Tests/SqlClientTests.cs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+                    .AddProcessorPipeline(p => p.AddProcessor(_ => spanProcessor.Object)))
+                .GetTracer(null);
+
+            using (new SqlClientCollector(tracer))
+            {
+                var operationId = Guid.NewGuid();
+                var sqlConnection = new SqlConnection(TestConnectionString);
+                var sqlCommand = sqlConnection.CreateCommand();
+                sqlCommand.CommandText = "SP_GetOrders";
+                sqlCommand.CommandType = CommandType.StoredProcedure;
```
Comment (cijothomas) [inline test/OpenTelemetry.Collector.Dependencies.Tests/SqlClientTests.cs]: could you add a test with CommandType as Text as well?

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                        if (span.IsRecording)
+                        {
+                            var dataSource = this.dataSourceFetcher.Fetch(connection);
+                            var commandText = this.commandTextFetcher.Fetch(command);
+
+                            span.PutComponentAttribute("sql");
+
+                            span.PutDatabaseTypeAttribute("sql");
+                            span.PutDatabaseInstanceAttribute((string)database);
+                            span.PutDatabaseStatementAttribute((string)commandText);
```
Comment (cijothomas) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: Seems like we read commandText always (for stored procedures as well as sql text), it'd be good to add comment here or todo here to sanitize the text (or make collection of sql text controlled by a setting in SqlClientCollectionOptions in the future.) as it can potentially contain secrets.

File: README.md [hunk 12]
```diff
@@ -57,7 +57,7 @@ Myget feeds:
...
 | ASP.NET Core                      | [![MyGet Nightly][OpenTelemetry-collect-aspnetcore-myget-image]][OpenTelemetry-collect-aspnetcore-myget-url]                   | [![NuGet Release][OpenTelemetry-collect-aspnetcore-nuget-image]][OpenTelemetry-collect-aspnetcore-nuget-url]                   |
-| .NET Core HttpClient & Azure SDKs | [![MyGet Nightly][OpenTelemetry-collect-deps-myget-image]][OpenTelemetry-collect-deps-myget-url]                               | [![NuGet Release][OpenTelemetry-collect-deps-nuget-image]][OpenTelemetry-collect-deps-nuget-url]                               |
+| .NET Core HttpClient, SqlClient, & Azure SDKs | [![MyGet Nightly][OpenTelemetry-collect-deps-myget-image]][OpenTelemetry-collect-deps-myget-url]                               | [![NuGet Release][OpenTelemetry-collect-deps-nuget-image]][OpenTelemetry-collect-deps-nuget-url]                               |
```
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply README.md]: Done

File: test/OpenTelemetry.Collector.Dependencies.Tests/SqlClientTests.cs (hunk 10 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Collector.Dependencies.Tests/SqlClientTests.cs]: Done

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs (hunk 11 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: @cijothomas Good call. I bit the bullet and added some basic options. Stored Procedure CommandText is captured by default. Text CommandText is NOT captured by default. Seemed like a much safer default behavior. Added `db.statementType` so you always get at least that. LMK your thoughts on all that.

File: src/OpenTelemetry.Collector.Dependencies/SqlClientCollectorOptions.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        /// Initializes a new instance of the <see cref="SqlClientCollectorOptions"/> class.
+        /// </summary>
+        public SqlClientCollectorOptions()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not the <see cref="SqlClientCollector"/> should capture the text of a <see cref="CommandType.StoredProcedure"/> command. Default value: True.
+        /// </summary>
+        public bool CaptureStoredProcedureCommandContent { get; set; } = true;
```
Comment (cijothomas) [inline src/OpenTelemetry.Collector.Dependencies/SqlClientCollectorOptions.cs]: Name of the Stored Procedure more suiting than Text of a StoredProcedure?

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs (hunk 11 above)
Comment (cijothomas) [reply src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: I agree this is right call - collect StoreProcedure name by default, collect SQL Command Text only if explicitly opted in via `SqlClientCollectionOptions`. 
Avoids privacy issues in the future.

File: src/OpenTelemetry.Collector.Dependencies/SqlClientCollectorOptions.cs (hunk 13 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Collector.Dependencies/SqlClientCollectorOptions.cs]: Updated

File: src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs (hunk 1 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs]: Yeah, we can leave it here.

File: src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+                            span.PutComponentAttribute("sql");
+
+                            span.PutDatabaseTypeAttribute("sql");
+                            span.PutPeerServiceAttribute((string)dataSource);
+                            span.PutDatabaseInstanceAttribute((string)database);
+
+                            if (this.commandTypeFetcher.Fetch(command) is CommandType commandType)
+                            {
+                                span.SetAttribute("db.statementType", commandType.ToString());
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: Could you make this a `cosnt string` please.

Comment (CodeBlanch) [reply src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs]: Done
